pkg/target/target.go

```diff
@@ -444,41 +459,83 @@ func convertToMatch(object map[string]interface{}) (*match.Match, error) {
 func (h *K8sValidationTarget) ToMatcher(u *unstructured.Unstructured) (constraints.Matcher, error) {
 	obj, found, err := unstructured.NestedMap(u.Object, "spec", "match")
 	if err != nil {
-		return nil, fmt.Errorf("%w: %v", ErrCreatingMather, err)
+		return nil, fmt.Errorf("%w: %v", ErrCreatingMatcher, err)
 	}
 	if found && obj != nil {
 		match, err := convertToMatch(obj)
 		if err != nil {
-			return nil, fmt.Errorf("%w: %v", ErrCreatingMather, err)
+			return nil, fmt.Errorf("%w: %v", ErrCreatingMatcher, err)
 		}
-		return &Matcher{match}, nil
+		return &Matcher{match, &h.cache}, nil
 	}
-	return nil, fmt.Errorf("%w: %s %s has no match found", ErrCreatingMather, u.GetKind(), u.GetName())
+	return nil, fmt.Errorf("%w: %s %s has no match found", ErrCreatingMatcher, u.GetKind(), u.GetName())
 }
 
 // Matcher implements constraint.Matcher.
-// TODO: add cache when handler.Cache interface is available.
 type Matcher struct {
 	match *match.Match
+	cache *nsCache
+}
+
+type nsCache struct {
+	lock  sync.RWMutex
+	cache map[string]*corev1.Namespace
+}
+
+func (nc *nsCache) Add(key string, ns *corev1.Namespace) error {
+	nc.lock.Lock()
+	defer nc.lock.Unlock()
+
+	if nc.cache == nil {
+		nc.cache = make(map[string]*corev1.Namespace)
+	}
+
+	nc.cache[key] = ns
+	return nil
+}
+
+func (nc *nsCache) Get(key string) (*corev1.Namespace, error) {
+	nc.lock.RLock()
+	defer nc.lock.RUnlock()
+
+	ns, ok := nc.cache[key]
+	if !ok {
+		return nil, nil
+	}
+	return ns, nil
+}
+
+func (nc *nsCache) Remove(key string) {
+	nc.lock.Lock()
+	defer nc.lock.Unlock()
+	delete(nc.cache, key)
 }
 
 func (m *Matcher) Match(review interface{}) (bool, error) {
+	var gkReq *gkReview
 	switch req := review.(type) {
 	case *gkReview:
-		obj, oldObj, ns, err := gkReviewToObject(req)
-		if err != nil {
-			return false, err
-		}
-		return matchAny(m, ns, &obj, &oldObj)
+		gkReq = req
 	case gkReview:
-		obj, oldObj, ns, err := gkReviewToObject(&req)
+		gkReq = &req
+	default:
+		return false, fmt.Errorf("%w: expect %T, got %T", ErrReviewFormat, gkReview{}, review)
+	}
+
+	obj, oldObj, ns, err := gkReviewToObject(gkReq)
+	if err != nil {
+		return false, err
+	}
+
+	if ns == nil {
+		cachedNs, err := m.cache.Get(obj.GetNamespace())
```

Comment:
`gkReq.Namespace`, in case namespace isn't populated on the object (e.g. `default` namespace)? context:

https://github.com/open-policy-agent/gatekeeper/blob/80d9993d44294e27809910a92fde521cb78886a1/pkg/webhook/mutation.go#L189-L193

Likely not significant here, but seems slightly safer.

Reply:
done